onnxruntime/core/providers/cpu/llm/attention_helper.h

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#pragma once
+#include "core/common/common.h"
+#include "core/providers/common.h"
+
+namespace onnxruntime {
+namespace attention_helper {
+
+enum AttentionMaskType {
+  MASK_NONE,  // No mask
+  MASK_BOOL,  // Boolean mask with shape [batch_size, sequence_length, total_sequence_length] or [batch_size, q_num_heads, q_sequence_length, total_sequence_length
+  MASK_ADD,   // Additive mask with shape [batch_size, sequence_length, total_sequence_length] or [batch_size, q_num_heads, q_sequence_length, total_sequence_length
+};
+
+enum AttentionType {
+  kInvalid = 0,               // Invalid attention type
+  kMultiHeadedAttention = 1,  // Multi-headed attention (MHA)
+  kGroupQueryAttention = 2,   // Group query attention (GQA)
+  kMultiQueryAttention = 3,   // Multi-query attention (MQA)
+};
+
+enum QKMatMulOutputMode {
+  kQK = 0,         // Output Q*K
+  kQKMask = 1,     // Output Q*K + Mask
+  kQKSoftCap = 2,  // Output SoftCap(Q*K + Mask)
+  kQKSoftMax = 3,  // Output SoftMax(SoftCap(Q*K + Mask))
+};
+
+// Parameters deduced from node attributes and inputs/outputs.
+struct AttentionParameters {
+  // Attribute from ONNX definition
+  bool is_causal;
+  int kv_num_heads;  // K.shape[1] or V.shape[1] (4D)
+  int q_num_heads;   // Q.shape[1] (4D)
+  float scale;
+  float softcap;
+  int softmax_precision;
+  QKMatMulOutputMode qk_matmul_output_mode;
+  AttentionMaskType mask_type;
+
+  // From shapes
+  int batch_size;            // Q.shape[0], K.shape[0], V.shape[0] (4D)
```

Comment:
Do we need to provide 3D note also? Otherwise, it seems a bit confusing (looks like we only allow 4D here).